pkg/pool/pool.go

```diff
@@ -401,3 +502,212 @@ func (p *pool) Introspect() IntrospectResponse {
 		CoolingResources: p.coolDownQueue,
 	}
 }
+
+// IsManagedResource checks if the IPv4 address is managed by the handler
+func (p *pool) IsManagedResource(resourceID string) bool {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	for _, resource := range p.usedResources {
+		if resource.ResourceID == resourceID {
+			return true
+		}
+	}
+	return false
+}
+
+// findMinGroup finds the resource group with the lowest number of resources, and returns the group id along with its count of resources
+func findMinGroup(resourceGroups map[string][]Resource) (group string, minCount int) {
+	for groupID, resources := range resourceGroups {
+		if resources == nil || len(resources) == 0 {
+			continue
+		}
+
+		// initialize return values as the first key-value pair; when lower count is found, record the group and minCount
+		if group == "" || len(resources) < minCount {
+			minCount = len(resources)
+			group = groupID
+		}
+	}
+
+	return
+}
+
+// assignResourceFromMinGroup assigns a resource from the resource group that has the lowest number of resources
+func (p *pool) assignResourceFromMinGroup() Resource {
+	groupID, resourceCount := findMinGroup(p.warmResources)
+	if groupID == "" || resourceCount == 0 {
+		return Resource{}
+	}
+
+	resource := p.warmResources[groupID][0]
+	if resourceCount == 1 {
+		p.warmResources[groupID] = nil
+	} else {
+		p.warmResources[groupID] = p.warmResources[groupID][1:]
+	}
+
+	return resource
+}
+
+// assignResourceFromAnyGroup assigns a resource from any resource group with at least one resource available
+func (p *pool) assignResourceFromAnyGroup() Resource {
+	resource := Resource{}
+	for groupID, resources := range p.warmResources {
+		if resources == nil || len(resources) <= 0 {
+			continue
+		}
+
+		resource = resources[0]
+		if len(resources) == 1 {
+			p.warmResources[groupID] = nil
+		} else {
+			p.warmResources[groupID] = resources[1:]
+		}
+		break
+	}
+
+	return resource
+}
+
+// getPDDeviation returns the deviation in number of IPv4 addresses when PD is enabled, taking into account of all PD configuration params
+func (p *pool) getPDDeviation() int {
```

Comment:
Please add unit test. It is hard to eye check for all these calculations.

Reply:
Added unit tests.